### PR TITLE
Fix malicious buffer deserialization panics.

### PR DESCRIPTION
**Pull request by vldm**
Closes https://github.com/exonum/exonum/issues/296

After some fuzz tests, and additional review I detected new panics in deserialization.

All test cases are filed in issue.

### CODE REVIEW

File: exonum/src/encoding/tests.rs [hunk 1]
```diff
@@ -24,22 +24,51 @@ use messages::{RawMessage, Message, FromRaw, Connect, Propose, Prevote, Precommi
 
 use super::{Field, Offset};
 
+#[allow(dead_code)]
+mod ignore_new {
```
Comment (alekseysidorov) [inline exonum/src/encoding/tests.rs]: What is purpose of this module?

File: exonum/src/messages/tests.rs [hunk 2]
```diff
@@ -25,3 +27,21 @@ fn test_message_without_fields() {
     }
     drop(NoFields::new(&SecretKey::new([1; 64])));
 }
+
+#[test]
+#[allow(dead_code)]
```
Comment (alekseysidorov) [inline exonum/src/messages/tests.rs]: Why you use this attribute? Can you explain me?

File: exonum/src/encoding/tests.rs (hunk 1 above)
Comment (vldm) [inline exonum/src/encoding/tests.rs]: I use `encoding_struct` to set buffer schema. So I need to ignore `dead_code` warning about `new` method. I add some comment.

File: exonum/src/messages/tests.rs (hunk 2 above)
Comment (vldm) [inline exonum/src/messages/tests.rs]: the same as above.

**Pull request merged**